mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala

```diff
@@ -74,26 +74,41 @@ class HashingTF @Since("1.4.0") (@Since("1.4.0") override val uid: String)
 
   setDefault(numFeatures -> (1 << 18), binary -> false)
 
+  private[this] var hashingTF = new feature.HashingTF($(numFeatures)).setBinary($(binary))
+
   /** @group getParam */
   @Since("1.2.0")
   def getNumFeatures: Int = $(numFeatures)
 
   /** @group setParam */
   @Since("1.2.0")
-  def setNumFeatures(value: Int): this.type = set(numFeatures, value)
+  def setNumFeatures(value: Int): this.type = {
+    val t = set(numFeatures, value)
+    hashingTF = new feature.HashingTF($(numFeatures)).setBinary($(binary))
+    t
+  }
 
   /** @group getParam */
   @Since("2.0.0")
   def getBinary: Boolean = $(binary)
 
   /** @group setParam */
   @Since("2.0.0")
-  def setBinary(value: Boolean): this.type = set(binary, value)
+  def setBinary(value: Boolean): this.type = {
+    val t = set(binary, value)
+    hashingTF.setBinary($(binary))
```

Comment:
I think we can't do other things except for set params in ```set***``` function, this is because PySpark ```set***``` doesn't call corresponding Scala function. In PySpark, we collect all params together and then pass them to Scala side, then call the following function:
```
def fit(dataset: Dataset[_], paramMap: ParamMap): M = {
    copy(paramMap).fit(dataset)
  }
```
which will skip the corresponding Scala ```set***``` functions. So it will make your change to old ```hashingTF``` doesn't take effect.